### PR TITLE
Add `separateDataFiles` to @rnw/cli and @rnw/codegen

### PR DESCRIPTION
**Pull request by ZihanChen-MSFT**
### Type of Change
- New feature (non-breaking change which adds functionality)

### Why
In devmain there are already requirements that, types involved in a turbo module needs to be part of the API.
Such project structure would be:
  - The API implemented in C++
  - Native code calls the API
  - The turbo module calls the API
  - JS code calls the turbo module

When a feature needs to be available to both C++ and JS, it is possible that the API and the turbo module will share some data structures. If by semantic they are always exactly the same, why not just publish generated types?

### What
In order to complete the story, the turbo module C++ codegen needs to generate a clean header files for custom types if required. Setting `codegenConfig.windows.separateDataTypes` does this.

## Testing
Tested in `sample-apps`

## Changelog
Should this change be included in the release notes: yes

`npx react-native codegen-windows` will generate two files if there is any custom type to generate. The additional file `NativeXXXDataTypes.g.h` contains all generated custom types.
 ###### Microsoft Reviewers: codeflow:open?pullrequest=https://github.com/microsoft/react-native-windows/pull/11990&drop=dogfoodAlpha

### CODE REVIEW

**Comment by acoates-ms** [comment]
Can we split the two files into separate generators, which would align with core where there are separate generators for the cpp vs h files.
Then instead of having to have another option for the config, we can just specify the generators we require.

File: packages/sample-apps/codegen/NativeMyModuleSpec.g.h [hunk 1]
```diff
@@ -8,18 +8,12 @@
  */
 #pragma once
 
+#include "NativeMyModuleTypes.g.h"
```
Comment (acoates-ms) [inline packages/sample-apps/codegen/NativeMyModuleSpec.g.h:11]: One use case I'm a little worried about is if in the future the API being called will want to add something to the struct that isn't currently available in the codegen.  If we put a #ifdef check around this include -- say `#ifndef NO_GENERATED_MYMODULE_TYPES`, then the codegen can also be used in a way that can just attach to existing structs, which would easily allow additional custom properties on the struct that are not available on the JS side.

**Comment by ZihanChen-MSFT** [comment]
> Can we split the two files into separate generators, which would align with core where there are separate generators for the cpp vs h files? Then instead of having to have another option for the config, we can just specify the generators we require.

Both file needs the input (in the intermediate format) to be compiled again to identify anonymous custom types before generating it, they shared many complex logic. I think it is good to just keep them in one generator.

File: packages/sample-apps/codegen/NativeMyModuleSpec.g.h (hunk 1 above)
Comment (ZihanChen-MSFT) [reply packages/sample-apps/codegen/NativeMyModuleSpec.g.h:11]: I can always skip this including

**Review by vmoroz — APPROVED** [review]
:shipit:

**Review by acoates-ms — APPROVED** [review]

**Pull request merged**